pypeit/core/flexure.py

```diff
@@ -55,68 +59,223 @@ def spat_flexure_shift(sciimg, slits, debug=False, maxlag = 20):
+    corr_sci = wvutils.get_xcorr_arc(sci_smash, cont_sub=True, percent_ceil=10., sigdetect=sigdetect, debug=debug)
+    corr_slits = wvutils.get_xcorr_arc(slits_smash, cont_sub=False, percent_ceil=None, input_thresh=1., debug=debug)
+
+    # run x-cross correlation
     lags, xcorr = utils.cross_correlate(corr_sci, corr_slits, maxlag)
     xcorr_denom = np.sqrt(np.sum(corr_sci*corr_sci)*np.sum(corr_slits*corr_slits))
     xcorr_norm = xcorr / xcorr_denom
-    # TODO -- Generate a QA plot
 
-    tampl_true, tampl, pix_max, twid, centerr, ww, arc_cont, nsig \
-            = arc.detect_lines(xcorr_norm, sigdetect=3.0, fit_frac_fwhm=1.5, fwhm=5.0,
-                               cont_frac_fwhm=1.0, cont_samp=30, nfind=1, debug=debug)
+    # detect the highest peak in the cross-correlation
+    _, _, pix_max, _, _, _, _, _ = arc.detect_lines(xcorr_norm, cont_subtract=False, input_thresh=0.,
+                                                    nfind=1, debug=debug)
     # No peak? -- e.g. data fills the entire detector
-    if len(tampl) == 0:
-        msgs.warn('No peak found in spatial flexure.  Assuming there is none...')
-        
+    if len(pix_max) == 0:
+        msgs.warn('No peak found in the x-correlation between the traced slits and the science/calib image.'
+                  '  Assuming there is NO SPATIAL FLEXURE.'+msgs.newline() + 'If a flexure is expected, '
+                  'consider either changing the maximum lag for the cross-correlation, '
+                  'or the "spat_flexure_sigdetect" parameter, or use the manual flexure correction.')
+
         return 0.
-    
-    # Find the peak
-    xcorr_max = np.interp(pix_max, np.arange(lags.shape[0]), xcorr_norm)
-    lag_max = np.interp(pix_max, np.arange(lags.shape[0]), lags)
-    msgs.info('Spatial flexure measured: {}'.format(lag_max[0]))
+    lag0 = np.where(lags == 0)[0][0]
+    shift = round(pix_max[0] - lag0, 3)
+    msgs.info('Spatial flexure measured: {}'.format(shift))
 
     if debug:
-        plt.figure(figsize=(14, 6))
-        plt.plot(lags, xcorr_norm, color='black', drawstyle='steps-mid', lw=3, label='x-corr')
-        plt.plot(lag_max[0], xcorr_max[0], 'g+', markersize=6.0, label='peak')
-        plt.title('Best shift = {:5.3f}'.format(lag_max[0]) + ',  corr_max = {:5.3f}'.format(xcorr_max[0]))
+        # 1D plot
+        xvals = np.arange(corr_slits.size)
+        plt.figure(figsize=(9, 8))
+        scale = np.nanmax(corr_slits) / np.nanmax(corr_sci)
+        plt.suptitle(f'Shift={shift:.1f} pixels', fontsize=18)
+        plt.plot(xvals, corr_slits, 'k', label='slitmask edge peaks')
+        plt.plot(xvals, np.roll(scale*corr_sci, -int(shift)), '--r', label='shifted science edge peaks')
         plt.legend()
+        plt.tight_layout()
         plt.show()
 
-    #tslits_shift = trace_slits.shift_slits(tslits_dict, lag_max)
-    # Now translate the tilts
+        # 2D plot
+        spat_flexure_qa(sciimg, slits, shift, gpm=np.logical_not(bpm), vrange=qa_vrange)
```

Comment:
remove this line?  it is duplicated below.
  or is the idea to generate 2 PNGs?

Reply:
No, this one is used only in debug mode and plot in a matplot window the whole image. No PNG is created. I uses the same script, but it does slightly different things.